presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +582,186 @@ private static NestedField nestedField(String name)
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
+                this.functionAndTypeManager = functionAndTypeManager;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
-                    call.getArguments().forEach(argument -> argument.accept(this, context));
+                ComplexTypeFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (functionResolution.isSubscriptFunction(call.getFunctionHandle()) ||
+                        functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName().equals(ELEMENT_AT)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (isPruningLambdaSubfieldsPossible(context)) {
+                        addRequiredLambdaSubfields(context, subfield);
+                        return null;
+                    }
+                    if (subfield.isPresent()) {
+                        context.subfields.add(subfield.get());
+                    }
+                    else {
+                        call.getArguments().forEach(argument -> argument.accept(this, context));
+                    }
                     return null;
                 }
+                if (isPushDownSubfieldsFromLambdasEnabled) {
+                    Optional<Set<Subfield>> lambdaSubfieldsOriginal = context.getLambdaSubfields();
+                    if ((functionDescriptor.isAccessingInputValues() && functionDescriptor.getLambdaDescriptors().isEmpty())
+                            || (functionDescriptor.isTransformingInput()) && !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
+                        // If function internally accesses the input values we cannot prune any unaccessed subfields since we do not know what subfields function accessed.
+                        // If function transforms the shape of the input map or array argument and the form of such transformation was not supplied in
+                        // getOutputToInputTransformationFunction() then we have to discard all collected accessed subfields in lambdas of outer calls because we do not know how
+                        // to attribute those subfields to the input argument.
+                        // Additionally, we have to stop collecting lambda subfields in both cases until we find another function that does not return input values.
+                        context.setLambdaSubfields(Optional.empty());
+                    }
+                    if (!functionDescriptor.isReturningInputValues()) {
+                        context.setLambdaSubfields(Optional.of(ImmutableSet.of()));
```

Comment:
why do we set this to an empty set?

Reply:
Actually, this triggers collecting of lambda subfield. I can add a comment or extract a function `startCollectingLambdaSubfields()` in order to make it more explicit.

Reply:
extract  startCollectingLambdaSubfields() function